### DISCUSSION TITLE
[Canvas] ItemControl do not set Canvas.Top or Canvas.Left

### DISCUSSION THREAD

**Discussion by cerobe** [AvaloniaUI/Avalonia#14296] [Q&A] [answered]
I'm using Avalonia 11.0.2 on desktop and trying to draw shapes on a Canvas using ItemsControl. **The problem is that the properties Canvas.Left and Canvas.Top are not being taken into account.**

**View:**
```xaml
<Canvas Width="{Binding Width}" Height="{Binding Height}" MaxWidth="200" MaxHeight="100" Background="Yellow" Margin="3">
    <Rectangle Fill="Blue" Height="10" Width="10" Canvas.Left="50" Canvas.Top="50"/>
    <ItemsControl ItemsSource="{Binding Tiles}" >
        <ItemsControl.ItemTemplate>
            <DataTemplate>
                <Rectangle Fill="Green" Height="{Binding Size}" Width="{Binding Size}" 
                           Canvas.Left="{Binding TopY}" Canvas.Top="{Binding TopY}"/>
            </DataTemplate>
        </ItemsControl.ItemTemplate>
    </ItemsControl>
</Canvas>
```
**ViewModel:**
```c#
public record CanvasTile(int Size, int TopX, int TopY);

public ObservableCollection<CanvasTile> Tiles { get; }

private const int TILE_SIZE = 10;
Tiles.Add(new CanvasTile(TILE_SIZE, 1 * TILE_SIZE, 1 * TILE_SIZE));
Tiles.Add(new CanvasTile(TILE_SIZE, 2 * TILE_SIZE, 2 * TILE_SIZE));
Tiles.Add(new CanvasTile(TILE_SIZE, 4 * TILE_SIZE, 4 * TILE_SIZE));
```

**Debug:**To show that the collection is set correctly
![nEnOZ](https://github.com/AvaloniaUI/Avalonia/assets/4987099/819bf546-4805-4e66-b2e1-98c8497cb556)

**Result:** As you can see below the green rectangles defined by the ItemControl are not at the right place. The blue rectangle is at the right place (I hardcoded his position outside of the (see code above)
![4tiKq](https://github.com/AvaloniaUI/Avalonia/assets/4987099/83370ecd-7734-4d96-a063-3c17b0bad6ba)

**Expected Result:** I've hardcoded the expected result to show what I should have get.
![Pv3ON](https://github.com/AvaloniaUI/Avalonia/assets/4987099/026eb23b-a262-49f0-8a8a-cf885a8ce649)

**Comment by thevortexcloud** [comment]
This is not right. Items controls have their own layout controls for laying out their items. By default it's a Vertical stack panel. You can override it by changing the `ItemPanel` template. You need to either change it to use a Canvas or put your items directly into the parent canvas.

**Reply by cerobe** [reply]
I 'm new Avalonia, I kinda understand your answer but do not know how to solve my problem using your suggestions. 

I can't put the items directly in the canvas as it's a dynamic list that will change with various event. That's why I created an observable collection in the ViewModel. 

You are talking about overidding ItemPanel. How should I do this?

I'm not precious about ItemControl. That's the only thing I found that seems to match my needs but If there is another way to do it, I'm listening.

**Reply by thevortexcloud** [reply]
>You are talking about overidding ItemPanel. How should I do this?

Something like this should work I think:

```xaml
<ItemsControl>
   <ItemsControl.ItemsPanel>
      <ItemsPanelTemplate>
         <Canvas/>
      </ItemsPanelTemplate>
   </ItemsControl.ItemsPanel>
</ItemsControl>
```

That said, is there a particular reason you need a Canvas? They are typically only used for fairly rare/special use cases.

**Reply by cerobe** [reply]
Thanks, I've tried this:

```xaml
<ItemsControl ItemsSource="{Binding Tiles}" >
	<ItemsControl.ItemsPanel>
		<ItemsPanelTemplate>
			<Canvas Width="{Binding Width}" Height="{Binding Height}" MaxWidth="200" MaxHeight="100" Background="Yellow" Margin="3"/>
		</ItemsPanelTemplate>
	</ItemsControl.ItemsPanel>
       <ItemsControl.ItemTemplate>
		<DataTemplate>
			<Rectangle Fill="Green" Height="{Binding Size}" Width="{Binding Size}" Canvas.Left="{Binding TopY}" Canvas.Top="{Binding TopY}"/>
		</DataTemplate>
	</ItemsControl.ItemTemplate>
</ItemsControl>
```

but it's giving me: (maybe I misunderstood your proposed solution?)
![image](https://github.com/AvaloniaUI/Avalonia/assets/4987099/cf8a5b7b-3f94-4dd9-a373-59930998f7d3)

I have view separated in 2 parts: a datagrid listing different 'map' and a viewer that is just a visual preview of the object in the grid. When I select an item in the datagrid I refresh the canvas and draw the 'map' in it. Is there a better way that using a canvas?

**Reply by timunie** [reply]
You will need a Style for ItemConatiner here

**Reply by cerobe** [reply]
Thanks @timunie  but how should I write this? looking at the [style doc](https://docs.avaloniaui.net/docs/basics/user-interface/styling/styles) I have put this together but it's clearly wrong as I have compiling error:

```xaml
<ItemsControl ItemsSource="{Binding Tiles}" >
	<ItemsControl.ItemsPanel>
		<ItemsPanelTemplate>
			<Canvas Width="{Binding Width}" Height="{Binding Height}" MaxWidth="200" MaxHeight="100" Background="Yellow" Margin="3"/>
		</ItemsPanelTemplate>
	</ItemsControl.ItemsPanel>
	<ItemsControl.ItemTemplate>
		<DataTemplate>
			<Rectangle Fill="Green" Height="{Binding Size}" Width="{Binding Size}"/>
		</DataTemplate>
	</ItemsControl.ItemTemplate>
	<ItemsControl.Styles>
		<Style Selector="Canvas">
			<Setter Property="Canvas.Left" Value="{Binding TopX}"/>
			<Setter Property="Canvas.Top" Value="{Binding TopY}"/>
		</Style>
	</ItemsControl.Styles>
</ItemsControl>
``` 
The error is:
![image](https://github.com/AvaloniaUI/Avalonia/assets/4987099/780f6b15-2229-41e6-b939-692d135f7069)

It seems that style does not know about the Item and cannot bind it's value?

**Reply by timunie** [reply]
Let me answer you next week. Have no computer at hand rn.

**Reply by timunie** [reply, marked as answer]
This fixes it for me
```xml
<ItemsControl.Styles>
	<Style Selector="ContentPresenter" x:DataType="vm:Tile">
		<Setter Property="Canvas.Left" Value="{Binding X}"/>
		<Setter Property="Canvas.Top" Value="{Binding Y}"/>
	</Style>
</ItemsControl.Styles>
```

Using DevTools [F12] you can see which control to target in your selector: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/74a3007c-98fb-4d5f-9d6e-c9e0c6671596)

**Reply by cerobe** [reply]
Hi @timunie your answer is indeed fixing my problem. Thanks for your help.

I would like to contribute the doc and add more example in the Canvas page (and maybe Style). That would have help me for sure and will probably help others. Should I make a PR with this example? (as it's more advanced usage of the canvas that would fit nicely with the already existing one)

**Reply by thevortexcloud** [reply]
I would say this belongs more on the items control pages. Since it really applies to any panel with attached properties.

**Reply by timunie** [reply]
Yes, that sounds like a good addition on ItemsControl page. Could also go into the DeepDrive section. Let me ask @MikeCodesDotNET what he thinks where to place it. Then a PR is more than welcome ❤️

**Reply by timunie** [reply]
@cerobe you could add a page under docs > DeepDrives and describe how to use a custom ItemsPanel For ItemsControl and friends. From the ItemsControl-page please add a link to that docs.

**Reply by cerobe** [reply]
@timunie , I saw you message after I have raised a PR with the new example under the ItemsControl. See https://github.com/AvaloniaUI/avalonia-docs/pull/327 

I'm happy to create a DeepDive page if it's needed, but is that complex enough for being call a DeepDive?

**Reply by timunie** [reply]
The thing is, that this concept also applies to ListBox and fiends, so it would be "wrong" to only show it for ItemsControl imo.

**Reply by cerobe** [reply]
I see. I'm new to Avalonia and using it build a small custom app for my own usage. I'm happy to contribute to this really cool project. I see that you have created an [issue on that subject](https://github.com/AvaloniaUI/avalonia-docs/issues/326.). Once my current PR is closed, I'll take it, that will be a good opportunity to learn a bit more about Avalonia and it will fix a gap in the doc.

**Reply by Garreye** [reply]
For anyone doing this with, I would recommended being more specific with the style selector:
```
<ItemsControl.Styles>
	<Style Selector="ItemsControl > ContentPresenter" x:DataType="vm:Tile">
		<Setter Property="Canvas.Left" Value="{Binding X}"/>
		<Setter Property="Canvas.Top" Value="{Binding Y}"/>
	</Style>
</ItemsControl.Styles>
```
Otherwise, child controls of the ItemsControl that use a ContentPresenter will throw an exception if they don't match the data type.
